### DISCUSSION TITLE
Make more functions static

### DISCUSSION THREAD

**Discussion by matthias-ronge** [kitodo/kitodo-production#5502] [General]
Production uses a service loader to initialize services once at runtime and to make them available later in the application. This makes sense for services that provide functionality from modules, since in this case the modules are loaded once at runtime. However, it makes no sense for static functions that are provided by services that do not use modules (and which do not have to be declared static).

**Goal:** In the services, declare all functions that work statically and do not use a module as `static`. Call these functions statically, without redirection via the service loader. If a service only provides static functions, remove the service from the service loader altogether.

**Comment by matthias-ronge** [comment]
There are a number of classes in Production called `Service`. Some of them are related to database objects. In this case, the services perform actions on the database objects. The functionality is in the service, the data in the data-management module.

    class                                 | service manager | module(s)
    --------------------------------------+-----------------+---------------------------------------
    AuthorityService                      | yes             | data-management
    BatchService                          | yes             | data-management
    ClientSearchDatabaseService           | no              |
    ClientSearchService                   | no              |
    ClientService                         | yes             | data-management
    CommentService                        | yes             | data-management
    DocketService                         | yes             | data-management
    FilterService                         | yes             | data-management
    FolderService                         | yes             | data-management
    LdapGroupService                      | yes             | data-management
    LdapServerService                     | yes             | data-management
    ListColumnService                     | yes             | data-management
    ProcessService                        | yes             | data-management, docket
    ProjectSearchService                  | no              |
    ProjectService                        | yes             | data-management
    PropertyService                       | yes             | data-management
    RoleService                           | yes             | data-management
    RulesetService                        | yes             | data-management
    SearchService                         | no              |
    TaskService                           | yes             | data-management
    TemplateService                       | yes             | data-management
    TitleSearchService                    | no              |
    UserService                           | yes             | data-management
    WorkflowConditionService              | yes             | data-management
    WorkflowService                       | yes             | data-management

Other services are related to the modules. The services call functions on the modules. The functionality is in the module, the services should do little except to access the module.

    class                                 | service manager | module(s)
    --------------------------------------+-----------------+---------------------------------------
    CommandService                        | yes             | command
    DataEditorService                     | yes             | dataeditor
    FileService                           | yes             | file-management
    ImageService                          | yes             | image-management
    ImportService                         | yes             | query-url-import, xml-schema-converter
    LongTermPreservationValidationService | yes             | longtimepreservationvalidation
    MetadataValidationService             | yes             | validation
    MetsService                           | yes             | dataformat
    RulesetManagementService              | yes             | dataeditor

Some services are neither-nor. There is simply functionality in the services. Basically they are utility classes. Many of them use the service loader, some don't. In addition, there are a number of utility classes with other names.

    class                                 | service manager | module(s)
    --------------------------------------+-----------------+---------------------------------------
    FileStructureValidationService        | yes             |
    IndexingService                       | yes             |
    MassImportService                     | yes             |
    MigrationService                      | yes             |
    SchemaService                         | yes             |
    SecurityAccessService                 | yes             |
    SessionService                        | yes             |
    CalendarService                       | no              |
    KitodoScriptService                   | no              |
    SubfolderFactoryService               | no              |
    WorkflowControllerService             | no              |
    
    BatchHelper                           | no              |
    BatchProcessHelper                    | no              |
    BatchTaskHelper                       | no              |
    DateUtils                             | no              |
    FacesUtils                            | no              |
    Helper                                | no              |
    HibernateUtil                         | no              |
    ImageHelper                           | no              |
    JaxbXmlUtils                          | no              |
    MetadataHelper                        | no              |
    WikiFieldHelper                       | no              |
    XMLUtils                              | no              |
    XsltHelper                            | no              |

**Comment by solth** [comment]
I converted the original issue into a discussion because this is a slightly controversial topic and I know that not everyone agrees on the general usage of the `static` modifier for methods in Kitodo.Production.

@henning-gerhardt please feel free to contribute to this discussion here!

**Reply by henning-gerhardt** [reply]
In my opinion there are only three reasons where static method should be used:
- to get a class instance (singleton)
- to initialize the class for a specific issue which could not be solved elsewhere in Java
- rules against the polymorph rules of oop

Maybe there a lot of other good reason not to use static methods but I don't think that I can convince someone in English nor in my native language and I did not discuss such specific details in a non-native language.

**Reply by markusweigelt** [reply]
I fully agree to that! So basically we should ask oneself: "Does it make sense to call this method, even if no object has been constructed yet?" (https://stackoverflow.com/questions/2671496/when-to-use-static-methods)

In principle, a lot of functionality can be packed into a static context, in order to be able to test it e.g. more simply, since one does not have to form an instance etc. first. However, this facets the source code and makes less use of inheritance and/or polymorphism. This in turn bloats the source code and makes it unnecessarily complicated. 

Therefore, we should ask ourselves the question more often than is currently done.
